<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Acquire-Data" data-toc-modified-id="Acquire-Data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Acquire Data</a></span><ul class="toc-item"><li><span><a href="#DataFrame-Metadata" data-toc-modified-id="DataFrame-Metadata-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>DataFrame Metadata</a></span><ul class="toc-item"><li><span><a href="#rainfall_summ-df" data-toc-modified-id="rainfall_summ-df-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>rainfall_summ df</a></span></li></ul></li><li><span><a href="#Explore-Unique-Values-by-Column" data-toc-modified-id="Explore-Unique-Values-by-Column-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Explore Unique Values by Column</a></span></li><li><span><a href="#Explore-Comment-Column" data-toc-modified-id="Explore-Comment-Column-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Explore Comment Column</a></span></li><li><span><a href="#Explore-Station-Id-Column" data-toc-modified-id="Explore-Station-Id-Column-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Explore Station Id Column</a></span></li></ul></li></ul></div>

In [56]:
import pandas as pd

## Acquire Data

We will acquire 3 sets of data from csv files. 

`water_quality` - BexarWaterQualityData.csv

`rainfall` - Rainfall24Hour.csv

`rainfall_summ` - RainfallSummaryData.csv

In [58]:
water_quality = 'BexarWaterQualityData.csv'
water_quality = pd.read_csv(water_quality)

rainfall = 'Rainfall24Hour.csv'
rainfall = pd.read_csv(rainfall)

rainfall_summ = 'RainfallSummaryData.csv'
rainfall_summ = pd.read_csv(rainfall_summ)

### DataFrame Metadata

We will look at summary data for each df.

In [59]:
water_quality.head()

station_id                      short_desc  \
0       12689  ROSILLO CREEK UPSTRM SALADO CR   
1       12689  ROSILLO CREEK UPSTRM SALADO CR   
2       12689  ROSILLO CREEK UPSTRM SALADO CR   
3       12689  ROSILLO CREEK UPSTRM SALADO CR   
4       12689  ROSILLO CREEK UPSTRM SALADO CR   

                                           long_desc   latitude  longitude  \
0  ROSILLO CREEK 0.1 KM ABOVE SALADO CREEK CONFLU...  29.320101 -98.406263   
1  ROSILLO CREEK 0.1 KM ABOVE SALADO CREEK CONFLU...  29.320101 -98.406263   
2  ROSILLO CREEK 0.1 KM ABOVE SALADO CREEK CONFLU...  29.320101 -98.406263   
3  ROSILLO CREEK 0.1 KM ABOVE SALADO CREEK CONFLU...  29.320101 -98.406263   
4  ROSILLO CREEK 0.1 KM ABOVE SALADO CREEK CONFLU...  29.320101 -98.406263   

     tag_id             end_date end_time  end_depth gtlt  ...  \
0  SA10917T  2008-09-18 00:00:00    16:52       0.28  NaN  ...   
1  SA10917T  2008-09-18 00:00:00    16:52       0.28  NaN  ...   
2  SA10917T  2008-09-18 00:00:00    16:52       0.28  NaN  ...   
3  SA10917T  2008-09-18 00:00:00    16:52       0.28  NaN  ...   
4  SA10917T  2008-09-18 00:00:00    16:52       0.28  NaN  ...   

                          common_name  common_units  \
0                   Water Temperature        &deg;C   
1                                 NaN           NaN   
2  Total Dissolved Solids, Calculated          mg/L   
3                                 NaN           NaN   
4                             E. coli    MPN/100 mL   

                                       detailed_desc      units  media  \
0  <p>Water temperature is an important factor in...      DEG C  Water   
1                                                NaN        CFS  Water   
2  <p>Total Dissolved Solids (TDS) is a measure o...      uS/cm  Water   
3                                                NaN         NS  Water   
4  <p>These bacteria are typically not harmful to...  MPN/100ml  Water   

            method min_value max_value county_name         comment  
0  N/A Calculation       NaN       NaN       Bexar  Abundant algae  
1  N/A Calculation       NaN       NaN       Bexar  Abundant algae  
2  N/A Calculation       NaN       NaN       Bexar  Abundant algae  
3  N/A Calculation       NaN       NaN       Bexar  Abundant algae  
4  N/A Calculation       NaN       NaN       Bexar  Abundant algae  

[5 rows x 23 columns]

In [60]:
water_quality.shape

(208373, 23)

In [61]:
water_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208373 entries, 0 to 208372
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   station_id     208373 non-null  int64  
 1   short_desc     208373 non-null  object 
 2   long_desc      208373 non-null  object 
 3   latitude       208373 non-null  float64
 4   longitude      208373 non-null  float64
 5   tag_id         208373 non-null  object 
 6   end_date       208373 non-null  object 
 7   end_time       208373 non-null  object 
 8   end_depth      208373 non-null  float64
 9   gtlt           19920 non-null   object 
 10  result         208373 non-null  object 
 11  parameter_cd   208373 non-null  int64  
 12  long_desc.1    208373 non-null  object 
 13  common_name    76444 non-null   object 
 14  common_units   76444 non-null   object 
 15  detailed_desc  76444 non-null   object 
 16  units          208373 non-null  object 
 17  media          208373 non-nul

In [62]:
rainfall.head()

name  latitude  longitude    datetime  24HourTotal
0  Calaveras 03  29.36972  -98.33231  2018-11-21          0.0
1  Calaveras 03  29.36972  -98.33231  2018-11-22          0.0
2  Calaveras 03  29.36972  -98.33231  2018-11-23          0.0
3  Calaveras 03  29.36972  -98.33231  2018-11-24          0.0
4  Calaveras 03  29.36972  -98.33231  2018-11-25          0.0

In [63]:
rainfall.shape

(17290, 5)

In [64]:
rainfall.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17290 entries, 0 to 17289
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   name         17290 non-null  object 
 1   latitude     17290 non-null  float64
 2   longitude    17290 non-null  float64
 3   datetime     17290 non-null  object 
 4   24HourTotal  17290 non-null  float64
dtypes: float64(3), object(2)
memory usage: 675.5+ KB


#### rainfall_summ df

- `location_name` possible aggregation level with `daily_rainfall_total_inches`.

- Possible time series analysis or aggregation by day or time range.

In [65]:
rainfall_summ.head()

location_name  latitude  longitude        date  \
0  Blanco Road Dam   29.6248  -98.52135  2018-01-01   
1  Blanco Road Dam   29.6248  -98.52135  2018-01-02   
2  Blanco Road Dam   29.6248  -98.52135  2018-01-03   
3  Blanco Road Dam   29.6248  -98.52135  2018-01-04   
4  Blanco Road Dam   29.6248  -98.52135  2018-01-05   

   daily_rainfall_total_inches  
0                         0.01  
1                         0.00  
2                         0.00  
3                         0.00  
4                         0.00

In [66]:
rainfall_summ.shape

(13395, 5)

In [67]:
rainfall_summ.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13395 entries, 0 to 13394
Data columns (total 5 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   location_name                13395 non-null  object 
 1   latitude                     13395 non-null  float64
 2   longitude                    13395 non-null  float64
 3   date                         13395 non-null  object 
 4   daily_rainfall_total_inches  13395 non-null  float64
dtypes: float64(3), object(2)
memory usage: 523.4+ KB


### Explore Unique Values by Column

In [21]:
for column in water_quality:
    print(f'There are {water_quality[column].nunique()} unique values in the {column} column')
    print()

There are 152 unique values in the station_id column

There are 151 unique values in the short_desc column

There are 152 unique values in the long_desc column

There are 150 unique values in the latitude column

There are 152 unique values in the longitude column

There are 11446 unique values in the tag_id column

There are 2434 unique values in the end_date column

There are 784 unique values in the end_time column

There are 167 unique values in the end_depth column

There are 2 unique values in the gtlt column

There are 5719 unique values in the result column

There are 318 unique values in the parameter_cd column

There are 318 unique values in the long_desc.1 column

There are 12 unique values in the common_name column

There are 5 unique values in the common_units column

There are 12 unique values in the detailed_desc column

There are 28 unique values in the units column

There are 3 unique values in the media column

There are 2 unique values in the method column

There are

### Explore Comment Column

In [32]:
water_quality.comment.value_counts()[water_quality.comment.value_counts() > 100]

Rec BacT                                                                                                                                   7095
WPP                                                                                                                                        6048
SARA SWQM WPP                                                                                                                              5901
Abundant algae                                                                                                                             3667
SARA-SWQM                                                                                                                                  2589
Stormwater event                                                                                                                           1802
CRP-SWQM                                                                                                                                

### Explore Station Id Column

There are 152 unique station ids in the `station_id` column. Possible level of aggregation.

In [37]:
water_quality.station_id.value_counts()

12897    10285
14256    10270
17066     8641
14198     7036
12870     6421
         ...  
18859       14
16582       11
12848       11
14225        9
15643        1
Name: station_id, Length: 152, dtype: int64

In [54]:
temp_df = water_quality.groupby(['station_id'])[['tag_id']].size().reset_index()
temp_df.rename({'0': 'tag_id'})

station_id     0
0         12689   219
1         12690   437
2         12692   247
3         12693   570
4         12698  1057
..          ...   ...
147       21547  1759
148       21599   948
149       21705   310
150       21825   429
151       22062    56

[152 rows x 2 columns]